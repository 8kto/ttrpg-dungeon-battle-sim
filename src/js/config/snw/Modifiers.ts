/**
 * @type {Record<number, StrengthModifierDef>}
 */
export const strengthModifiers = {
  4: {
    ToHit: -2,
    Damage: -1,
    Doors: '1',
    Carry: -10,
  },
  6: {
    ToHit: -1,
    Damage: 0,
    Doors: '1',
    Carry: -5,
  },
  8: {
    ToHit: 0,
    Damage: 0,
    Doors: '1-2',
    Carry: 0,
  },
  12: {
    ToHit: 0,
    Damage: 0,
    Doors: '1-2',
    Carry: 5,
  },
  15: {
    ToHit: 1,
    Damage: 0,
    Doors: '1-2',
    Carry: 10,
  },
  16: {
    ToHit: 1,
    Damage: 1,
    Doors: '1-3',
    Carry: 15,
  },
  17: {
    ToHit: 2,
    Damage: 2,
    Doors: '1-4',
    Carry: 30,
  },
  18: {
    ToHit: 2,
    Damage: 3,
    Doors: '1-5',
    Carry: 50,
  },
}

/**
 * @typedef DexterityModifierDef
 * @property {number} MissilesToHit
 * @property {number} ArmorClass
 */

/**
 * @type {Record<number, DexterityModifierDef>}
 */
export const dexterityModifiers = {
  8: {
    MissilesToHit: -1,
    ArmorClass: -1,
  },
  12: {
    MissilesToHit: 0,
    ArmorClass: 0,
  },
  18: {
    MissilesToHit: 1,
    ArmorClass: 1,
  },
}

/**
 * @typedef ConstitutionModifierDef
 * @property {number} HitPoints
 * @property {string} RaiseDeadSurvivalChance
 */

/**
 * @type {Record<number, ConstitutionModifierDef>}
 */
export const constitutionModifiers = {
  8: {
    HitPoints: -1,
    RaiseDeadSurvivalChance: '50%',
  },
  12: {
    HitPoints: 0,
    RaiseDeadSurvivalChance: '75%',
  },
  18: {
    HitPoints: 1,
    RaiseDeadSurvivalChance: '100%',
  },
}

/**
 * @typedef IntelligenceModifierDef
 * @property {number} MaxAdditionalLanguages
 * @property {number} MaxSpellLevel
 * @property {string} NewSpellUnderstandingChance
 * @property {string} SpellsPerLevel
 */

/**
 * @type {Record<number, IntelligenceModifierDef>}
 */
export const intelligenceModifiers = {
  7: {
    MaxAdditionalLanguages: 0,
    MaxSpellLevel: 4,
    NewSpellUnderstandingChance: '30%',
    SpellsPerLevel: '2/4',
  },
  8: {
    MaxAdditionalLanguages: 1,
    MaxSpellLevel: 5,
    NewSpellUnderstandingChance: '40%',
    SpellsPerLevel: '3/5',
  },
  9: {
    MaxAdditionalLanguages: 1,
    MaxSpellLevel: 5,
    NewSpellUnderstandingChance: '45%',
    SpellsPerLevel: '3/5',
  },
  10: {
    MaxAdditionalLanguages: 2,
    MaxSpellLevel: 5,
    NewSpellUnderstandingChance: '50%',
    SpellsPerLevel: '4/6',
  },
  11: {
    MaxAdditionalLanguages: 2,
    MaxSpellLevel: 6,
    NewSpellUnderstandingChance: '50%',
    SpellsPerLevel: '4/6',
  },
  12: {
    MaxAdditionalLanguages: 3,
    MaxSpellLevel: 6,
    NewSpellUnderstandingChance: '55%',
    SpellsPerLevel: '4/6',
  },
  13: {
    MaxAdditionalLanguages: 3,
    MaxSpellLevel: 7,
    NewSpellUnderstandingChance: '65%',
    SpellsPerLevel: '5/8',
  },
  14: {
    MaxAdditionalLanguages: 4,
    MaxSpellLevel: 7,
    NewSpellUnderstandingChance: '65%',
    SpellsPerLevel: '5/8',
  },
  15: {
    MaxAdditionalLanguages: 4,
    MaxSpellLevel: 8,
    NewSpellUnderstandingChance: '75%',
    SpellsPerLevel: '6/10',
  },
  16: {
    MaxAdditionalLanguages: 5,
    MaxSpellLevel: 8,
    NewSpellUnderstandingChance: '75%',
    SpellsPerLevel: '6/10',
  },
  17: {
    MaxAdditionalLanguages: 5,
    MaxSpellLevel: 9,
    NewSpellUnderstandingChance: '85%',
    SpellsPerLevel: '7/All',
  },
  18: {
    MaxAdditionalLanguages: 6,
    MaxSpellLevel: 9,
    NewSpellUnderstandingChance: '95%',
    SpellsPerLevel: '8/All',
  },
}

/**
 * @typedef CharismaModifierDef
 * @property {number} MaxNumberOfSpecialHirelings
 */

/**
 * @type {Record<number, CharismaModifierDef>}
 */
export const charismaModifiers = {
  4: { MaxNumberOfSpecialHirelings: 1 },
  6: { MaxNumberOfSpecialHirelings: 2 },
  8: { MaxNumberOfSpecialHirelings: 3 },
  12: { MaxNumberOfSpecialHirelings: 4 },
  15: { MaxNumberOfSpecialHirelings: 5 },
  17: { MaxNumberOfSpecialHirelings: 6 },
  18: { MaxNumberOfSpecialHirelings: 7 },
}
