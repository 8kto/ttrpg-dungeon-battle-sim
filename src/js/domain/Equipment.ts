export type EquipItem = {
  name: string
  weight: number
  cost: number
  ascArmorClass?: number
  damage?: string
  flags?: InventoryItemFlag
}

/**
 * Flags for special properties of inventory items.
 */
export enum InventoryItemFlag {
  TWO_HANDED = 0b000001,
  VAR_HANDED = 0b000010,
  MELEE_AND_MISSILE = 0b000100,
  TYPE_MELEE = 0b001000,
  TYPE_MISSILE = 0b010000,
}

export type EquipSet = {
  name: string
  items: Array<{ name: string; quantity: number }>
}
