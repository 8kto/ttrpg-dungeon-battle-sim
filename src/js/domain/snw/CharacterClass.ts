import type { Dice } from 'ttrpg-lib-dice'
import type { SavingThrow } from './SavingThrow'

export enum CharacterClass {
  Assassin = 'Assassin',
  Cleric = 'Cleric',
  Druid = 'Druid',
  Fighter = 'Fighter',
  MagicUser = 'MagicUser',
  Monk = 'Monk',
  Paladin = 'Paladin',
  Ranger = 'Ranger',
  Thief = 'Thief',
}

// TODO ancestries
// TODO ancestry to class map
export enum CharacterRace {
  Human = 'Human',
  Dwarf = 'Dwarf',
  Elf = 'Elf',
  HalfElf = 'HalfElf',
  Hobbit = 'Hobbit',
}

export enum AttrScore {
  Strength = 'Strength',
  Dexterity = 'Dexterity',
  Constitution = 'Constitution',
  Intelligence = 'Intelligence',
  Wisdom = 'Wisdom',
  Charisma = 'Charisma',
}

export enum Alignment {
  Chaotic = 'Chaotic',
  Neutral = 'Neutral',
  Lawful = 'Lawful',
}

export type PrimeAttribute = [AttrScore, number]

export interface CharacterClassDef {
  name: CharacterClass
  PrimeAttr: PrimeAttribute[]
  StrictAttr?: PrimeAttribute[]
  HitDice: Dice
  ArmorPermitted: string
  WeaponsPermitted: string
  Race: CharacterRace[]
  Alignment: Alignment[]
  SavingThrow: SavingThrow // TODO should depend on level
  $isCaster?: boolean
  $spellsAtTheFirstLevel?: number
  $meta?: string
}
