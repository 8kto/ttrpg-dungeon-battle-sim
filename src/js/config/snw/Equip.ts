import { EquipItem, InventoryItemFlag } from '../../domain/Equipment'

export const Armor: Array<EquipItem> = [
  { cost: 15, name: 'Shield', weight: 10, armorClass: '+1 [-1]' },
  { cost: 5, name: 'Leather', weight: 25, armorClass: '+2 [-2]' },
  { cost: 30, name: 'Ring', weight: 40, armorClass: '+3 [-3]' },
  { cost: 75, name: 'Chain', weight: 50, armorClass: '+4 [-4]' },
  { cost: 100, name: 'Plate', weight: 70, armorClass: '+6 [-6]' },
]

export const Weapons: Array<EquipItem> = [
  { cost: 2, name: 'Arrows (20)', weight: 1, damage: 'd6' },
  { cost: 2, name: 'Bolts, heavy (20)', weight: 1, damage: 'd6+1' },
  { cost: 2, name: 'Bolts, light (20)', weight: 1, damage: 'd4+1' },
  { cost: 5, name: 'Axe, Battle', weight: 15, flags: InventoryItemFlag.VAR_HANDED | InventoryItemFlag.TWO_HANDED, damage: 'd8' },
  { cost: 1, name: 'Axe, Hand', weight: 5, flags: InventoryItemFlag.MELEE_AND_MISSILE, damage: 'd6' },
  { cost: 60, name: 'Bow, Long', weight: 5, damage: '' },
  { cost: 15, name: 'Bow, Short', weight: 5, damage: '' },
  { cost: 0, name: 'Club', weight: 10, damage: 'd4' },
  { cost: 20, name: 'Crossbow, Heavy', weight: 5, damage: '' },
  { cost: 12, name: 'Crossbow, Light', weight: 5, damage: '' },
  { cost: 2, name: 'Dagger', weight: 2, flags: InventoryItemFlag.MELEE_AND_MISSILE, damage: 'd4' },
  { cost: 0.2, name: 'Dart', weight: 1, damage: 'd3' },
  { cost: 0.5, name: 'Javelin', weight: 5, damage: 'd6' },
  { cost: 10, name: 'Mace', weight: 10, damage: 'd6' },
  { cost: 1, name: 'Sling', weight: 0, damage: 'd4' },
  { cost: 0, name: 'Sling Stones (20)', weight: 5, damage: 'd4' },
  { cost: 1, name: 'Spear', weight: 10, flags: InventoryItemFlag.VAR_HANDED | InventoryItemFlag.MELEE_AND_MISSILE | InventoryItemFlag.TWO_HANDED, damage: 'd6' },
  { cost: 0, name: 'Staff', weight: 10, damage: 'd6' },
  { cost: 8, name: 'Sword, Short', weight: 5, damage: 'd6' },
  { cost: 20, name: 'Sword, Bastard', weight: 10, flags: InventoryItemFlag.VAR_HANDED | InventoryItemFlag.TWO_HANDED, damage: 'd8' },
  { cost: 15, name: 'Sword, Long', weight: 10, damage: 'd8' },
  { cost: 30, name: 'Sword, Two-Handed', weight: 15, damage: 'd10' },
  { cost: 1, name: 'Hammer, war', weight: 10, damage: 'd4+1' },
]

export const Equip: Array<EquipItem> = [
  { cost: 0, name: 'Basic accessories', weight: 8 },
  { cost: 5, name: 'Backpack', weight: 2 },
  { cost: 2, name: 'Barrel', weight: 25 },
  { cost: 1, name: 'Bedroll', weight: 5 },
  { cost: 1, name: 'Bell', weight: 0 },
  { cost: 5, name: 'Block and tackle', weight: 2 },
  { cost: 2, name: 'Bottle of wine, glass', weight: 6 },
  { cost: 0.01, name: 'Candle', weight: 0.5 },
  { cost: 0.1, name: 'Canvas (per square yard)', weight: 0.5 },
  { cost: 1, name: 'Case (map or scroll)', weight: 1 },
  { cost: 30, name: 'Chain (10 feet)', weight: 50 },
  { cost: 0.05, name: 'Chalk, 1 piece', weight: 0 },
  { cost: 2, name: 'Chest', weight: 25 },
  { cost: 0.2, name: 'Crowbar', weight: 10 },
  { cost: 4, name: 'Fishing net (25 square feet)', weight: 5 },
  { cost: 0.03, name: 'Flask, leather', weight: 0.5 },
  { cost: 1, name: 'Flint & steel', weight: 1 },
  { cost: 10, name: 'Garlic, charmed', weight: 0 },
  { cost: 1, name: 'Grappling Hook', weight: 4 },
  { cost: 0.5, name: 'Hammer', weight: 2 },
  { cost: 1, name: 'Holy symbol, wooden', weight: 0 },
  { cost: 25, name: 'Holy symbol, silver', weight: 0.5 },
  { cost: 25, name: 'Holy water (flask)', weight: 1 },
  { cost: 1, name: 'Ink (1-ounce bottle)', weight: 0.1 },
  { cost: 0.05, name: 'Ladder (10 foot)', weight: 20 },
  { cost: 0.1, name: 'Lamp, bronze', weight: 2 },
  { cost: 12, name: 'Lantern, bullseye', weight: 3 },
  { cost: 7, name: 'Lantern, hooded', weight: 2 },
  { cost: 20, name: 'Lock', weight: 1 },
  { cost: 15, name: 'Manacles', weight: 5 },
  { cost: 20, name: 'Mirror', weight: 0.5 },
  { cost: 5, name: 'Musical instrument', weight: 3 },
  { cost: 0.1, name: 'Oil, lamp (1 pint)', weight: 1 },
  { cost: 0.2, name: 'Parchment (sheet)', weight: 0.5 },
  { cost: 0.2, name: 'Pole (10 foot)', weight: 10 },
  { cost: 0.5, name: 'Pot, iron', weight: 5 },
  { cost: 1, name: 'Rations, trail', weight: 2 },
  { cost: 0.5, name: 'Rations, dried', weight: 2 },
  { cost: 1, name: 'Rope, hemp (50 feet)', weight: 5 },
  { cost: 10, name: 'Rope, silk (50 feet)', weight: 2 },
  { cost: 1, name: 'Sack (15 pounds capacity)', weight: 1 },
  { cost: 2, name: 'Sack, (30 pounds capacity)', weight: 2 },
  { cost: 2, name: 'Shovel', weight: 5 },
  { cost: 0.5, name: 'Signal whistle', weight: 0 },
  { cost: 25, name: 'Spellbook, blank', weight: 1 },
  { cost: 0.05, name: 'Spike, iron', weight: 2 },
  { cost: 10, name: 'Tent', weight: 20 },
  { cost: 0.01, name: 'Torch', weight: 1 },
  { cost: 1, name: 'Waterskin', weight: 2 },
  { cost: 0.1, name: 'Wolfsbane', weight: 0.5 },
]

export const AllEquipment: Array<EquipItem> = [...Armor, ...Weapons, ...Equip]
