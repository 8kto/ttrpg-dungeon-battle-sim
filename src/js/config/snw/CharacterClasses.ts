import { Dice } from '../../domain/Dice'
import { Alignment, AttrScore, CharacterClass, CharacterClassDef, CharacterRace } from '../../domain/snw/CharacterClass'

export const ANY_RACE = Object.values(CharacterRace) as CharacterRace[]

export const ANY_ALIGNMENT = Object.values(Alignment) as Alignment[]

export const PRIME_ATTR_MIN = 13

const AssassinClassDef: CharacterClassDef = {
  name: CharacterClass.Assassin,
  PrimeAttr: [
    [AttrScore.Strength, PRIME_ATTR_MIN],
    [AttrScore.Dexterity, PRIME_ATTR_MIN],
    [AttrScore.Intelligence, PRIME_ATTR_MIN],
  ],
  StrictAttr: [
    [AttrScore.Strength, 12],
    [AttrScore.Dexterity, 12],
    [AttrScore.Intelligence, 12],
  ],
  HitDice: Dice.d6,
  ArmorPermitted: 'Leather armor only, shield permitted',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Neutral, Alignment.Chaotic],
}

const ClericClassDef: CharacterClassDef = {
  name: CharacterClass.Cleric,
  PrimeAttr: [[AttrScore.Wisdom, PRIME_ATTR_MIN]],
  HitDice: Dice.d6,
  ArmorPermitted: 'Any',
  WeaponsPermitted:
    'Blunt weapons only (club, flail, hammer, mace, staff, etc.). ' +
    'No missile weapons other than oil or slings if the Referee permits.',
  Race: [CharacterRace.Human, CharacterRace.HalfElf],
  Alignment: [Alignment.Lawful, Alignment.Chaotic],
  $isCaster: true,
  $spellsAtTheFirstLevel: 0,
}

const DruidClassDef: CharacterClassDef = {
  name: CharacterClass.Druid,
  PrimeAttr: [
    [AttrScore.Wisdom, PRIME_ATTR_MIN],
    [AttrScore.Charisma, PRIME_ATTR_MIN],
  ],
  StrictAttr: [
    [AttrScore.Wisdom, 12],
    [AttrScore.Charisma, 14],
  ],
  HitDice: Dice.d6,
  ArmorPermitted: 'Leather armor, wooden shield',
  WeaponsPermitted: 'Dagger, sickle-shaped sword (treat as short sword), spear, sling, oil',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Neutral],
  $isCaster: true,
  $spellsAtTheFirstLevel: 1,
}

const FighterClassDef: CharacterClassDef = {
  name: CharacterClass.Fighter,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: ANY_RACE,
  Alignment: ANY_ALIGNMENT,
}

const MagicUserClassDef: CharacterClassDef = {
  name: CharacterClass.MagicUser,
  PrimeAttr: [[AttrScore.Intelligence, PRIME_ATTR_MIN]],
  HitDice: Dice.d4,
  ArmorPermitted: 'None',
  WeaponsPermitted: 'Dagger, staff, and darts',
  Race: [CharacterRace.Elf, CharacterRace.HalfElf, CharacterRace.Human],
  Alignment: ANY_ALIGNMENT,
  $isCaster: true,
  $spellsAtTheFirstLevel: 1,
}

const MonkClassDef: CharacterClassDef = {
  name: CharacterClass.Monk,
  PrimeAttr: [[AttrScore.Wisdom, PRIME_ATTR_MIN]],
  StrictAttr: [
    [AttrScore.Strength, 12],
    [AttrScore.Dexterity, 15],
    [AttrScore.Wisdom, 15],
  ],
  HitDice: Dice.d4,
  ArmorPermitted: 'None',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: ANY_ALIGNMENT,
}

const PaladinClassDef: CharacterClassDef = {
  name: CharacterClass.Paladin,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  StrictAttr: [[AttrScore.Charisma, 17]],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Lawful],
}

const RangerClassDef: CharacterClassDef = {
  name: CharacterClass.Ranger,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  StrictAttr: [
    [AttrScore.Constitution, 15],
    [AttrScore.Intelligence, 12],
    [AttrScore.Wisdom, 12],
  ],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Lawful],
  $meta: '2d8 at first level, 1d8/level thereafter',
}

const ThiefClassDef: CharacterClassDef = {
  name: CharacterClass.Thief,
  PrimeAttr: [[AttrScore.Dexterity, PRIME_ATTR_MIN]],
  HitDice: Dice.d4,
  ArmorPermitted: 'Leather armor only; no shield',
  WeaponsPermitted: 'All, but magical weapons are limited to daggers and swords',
  Race: ANY_RACE,
  Alignment: [Alignment.Neutral, Alignment.Chaotic],
}

/**
 * Record for Character Class Definitions
 */
export const CharacterClasses: Record<CharacterClass, CharacterClassDef> = {
  [CharacterClass.Assassin]: AssassinClassDef,
  [CharacterClass.Cleric]: ClericClassDef,
  [CharacterClass.Druid]: DruidClassDef,
  [CharacterClass.Fighter]: FighterClassDef,
  [CharacterClass.MagicUser]: MagicUserClassDef,
  [CharacterClass.Monk]: MonkClassDef,
  [CharacterClass.Paladin]: PaladinClassDef,
  [CharacterClass.Ranger]: RangerClassDef,
  [CharacterClass.Thief]: ThiefClassDef,
}
