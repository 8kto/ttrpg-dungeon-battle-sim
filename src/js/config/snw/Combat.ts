import { CharacterClass } from '../../domain/snw/CharacterClass'

/**
 * See the Quick Alternative rules in S&W rulebook: no attack tables,
 * just plain d20 and progression with levels.
 *
 * NB: All classes have Base To-Hit = 0 up to 3rd level.
 * NB: Note: I didn't include any logic for levels higher than the first.
 * It is intended to be a simple random character generator, after all.
 */
export const QuickAscendingArmorClassBaseToHit: Record<number, Record<CharacterClass, number>> = {
  1: {
    [CharacterClass.Assassin]: 0,
    [CharacterClass.Cleric]: 0,
    [CharacterClass.Druid]: 0,
    [CharacterClass.Fighter]: 0,
    [CharacterClass.MagicUser]: 0,
    [CharacterClass.Monk]: 0,
    [CharacterClass.Paladin]: 0,
    [CharacterClass.Ranger]: 0,
    [CharacterClass.Thief]: 0,
  },
  // Below are not used ATM, since the char generator supports only 1st level chars
  2: {
    [CharacterClass.Assassin]: 0,
    [CharacterClass.Cleric]: 0,
    [CharacterClass.Druid]: 0,
    [CharacterClass.Fighter]: 0,
    [CharacterClass.MagicUser]: 0,
    [CharacterClass.Monk]: 0,
    [CharacterClass.Paladin]: 0,
    [CharacterClass.Ranger]: 0,
    [CharacterClass.Thief]: 0,
  },
  3: {
    [CharacterClass.Assassin]: 0,
    [CharacterClass.Cleric]: 1,
    [CharacterClass.Druid]: 1,
    [CharacterClass.Fighter]: 1,
    [CharacterClass.MagicUser]: 0,
    [CharacterClass.Monk]: 1,
    [CharacterClass.Paladin]: 1,
    [CharacterClass.Ranger]: 1,
    [CharacterClass.Thief]: 0,
  },
}
