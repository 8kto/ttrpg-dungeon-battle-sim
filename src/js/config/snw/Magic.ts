import { Spell } from '../../domain/snw/Magic'

export const MagicUserSpells: Record<string, Spell> = {
  'Charm Person': {
    name: 'Charm Person',
    level: 1,
  },
  'Detect Magic': {
    name: 'Detect Magic',
    level: 1,
  },
  'Hold Portal': {
    name: 'Hold Portal',
    level: 1,
  },
  Light: {
    name: 'Light',
    level: 1,
  },
  'Magic Missile': {
    name: 'Magic Missile',
    level: 1,
  },
  'Protection from Evil': {
    name: 'Protection from Evil',
    level: 1,
  },
  'Read Languages': {
    name: 'Read Languages',
    level: 1,
  },
  'Read Magic': {
    name: 'Read Magic',
    level: 1,
  },
  Shield: {
    name: 'Shield',
    level: 1,
  },
  Sleep: {
    name: 'Sleep',
    level: 1,
  },
}

export const ClericSpells: Record<string, Spell> = {
  'Cure Light Wounds': {
    name: 'Cure Light Wounds',
    level: 1,
  },
  'Detect Evil': {
    name: 'Detect Evil',
    level: 1,
  },
  'Detect Magic': {
    name: 'Detect Magic',
    level: 1,
  },
  Light: {
    name: 'Light',
    level: 1,
  },
  'Protection from Evil': {
    name: 'Protection from Evil',
    level: 1,
  },
  'Purify Food and Drink': {
    name: 'Purify Food and Drink',
    level: 1,
  },
}

export const DruidSpells: Record<string, Spell> = {
  'Detect Magic': {
    name: 'Detect Magic',
    level: 1,
  },
  'Detect Snares & Pits': {
    name: 'Detect Snares & Pits',
    level: 1,
  },
  'Faerie Fire': {
    name: 'Faerie Fire',
    level: 1,
  },
  'Locate Animals': {
    name: 'Locate Animals',
    level: 1,
  },
  'Predict Weather': {
    name: 'Predict Weather',
    level: 1,
  },
  'Purify Water': {
    name: 'Purify Water',
    level: 1,
  },
}
