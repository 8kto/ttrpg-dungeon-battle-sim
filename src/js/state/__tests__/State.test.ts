import { EquipItem } from '../../domain/Equipment'
import { CharacterClass } from '../../domain/snw/CharacterClass'
import {
  CharacterOptions,
  DEFAULT_INVENTORY,
  DEFAULT_INVENTORY_ID,
  DEFAULT_INVENTORY_ITEMS,
  getState,
  State,
} from '../State'

describe('State', () => {
  let state: State
  // @ts-ignore
  jest.SpyInstance = jest.spyOn(console, 'error').mockImplementation(() => {
    void 0
  })

  beforeEach(() => {
    localStorage.clear()
    State.resetInstance()
    state = getState()
  })

  describe('default state', () => {
    it('should initialize with default inventory', () => {
      const inventories = state.getInventories()
      expect(inventories).toMatchSnapshot()
    })

    it('should return current inventory ID', () => {
      expect(state.getCurrentInventoryId()).toEqual(DEFAULT_INVENTORY_ID)
    })

    it('should serialize and deserialize inventories', () => {
      state.serialize()
      const serializedInventories = state.getSerializedInventories()

      expect(serializedInventories).toEqual({
        [DEFAULT_INVENTORY_ID]: DEFAULT_INVENTORY,
      })
    })
  })

  describe('inventory items', () => {
    it('should add inventory items', () => {
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 10, name: 'Sword', weight: 5 })
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 10, name: 'Sword', weight: 5 })
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 70, name: 'Plate', weight: 100 })

      const inventory = state.getInventory(DEFAULT_INVENTORY_ID)
      expect(inventory.items).toEqual({
        'Basic accessories': {
          cost: 0,
          name: 'Basic accessories',
          quantity: 1,
          weight: 8,
        },
        Plate: {
          cost: 70,
          name: 'Plate',
          quantity: 1,
          weight: 100,
        },
        Sword: {
          cost: 10,
          name: 'Sword',
          quantity: 2,
          weight: 5,
        },
      })
    })

    it('should remove item from inventory', () => {
      const item: EquipItem = { cost: 10, name: 'Mace', weight: 5 }
      state.addToInventory(DEFAULT_INVENTORY_ID, item)
      state.removeFromInventory(DEFAULT_INVENTORY_ID, 'Mace')
      const inventory = state.getInventory(DEFAULT_INVENTORY_ID)

      expect(inventory.items['Mace']).toBeUndefined()
    })

    it('should set and get current inventory ID', () => {
      const newInventoryId = 'newId'
      state.setInventory(newInventoryId, DEFAULT_INVENTORY)
      state.setCurrentInventoryId(newInventoryId)
      expect(state.getCurrentInventoryId()).toBe(newInventoryId)
    })

    it('should throw for unknown inventory ID', () => {
      const newInventoryId = 'newId'

      expect(() => state.setCurrentInventoryId(newInventoryId)).toThrow(`Inventory does not exist: ${newInventoryId}`)
    })

    it('should reset inventory items', () => {
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 10, name: 'Sword', weight: 5 })
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 10, name: 'Sword', weight: 5 })
      state.addToInventory(DEFAULT_INVENTORY_ID, { cost: 70, name: 'Plate', weight: 100 })

      state.resetInventoryItems(DEFAULT_INVENTORY_ID)

      const inventory = state.getInventory(DEFAULT_INVENTORY_ID)
      expect(inventory.items).toEqual(DEFAULT_INVENTORY_ITEMS)
    })
  })

  describe('character', () => {
    it('should set and remove character in inventory', () => {
      const characterOptions: CharacterOptions = {
        characterClass: CharacterClass.MagicUser,
        spells: {
          'Charm Person': {
            level: 1,
            name: 'Charm Person',
          },
          'Detect Magic': {
            level: 1,
            name: 'Detect Magic',
          },
          'Hold Portal': {
            level: 1,
            name: 'Hold Portal',
          },
          Light: {
            level: 1,
            name: 'Light',
          },
          'Protection from Evil': {
            level: 1,
            name: 'Protection from Evil',
          },
          'Read Languages': {
            level: 1,
            name: 'Read Languages',
          },
          Shield: {
            level: 1,
            name: 'Shield',
          },
          Sleep: {
            level: 1,
            name: 'Sleep',
          },
        },
        stats: {
          Charisma: {
            MaxNumberOfSpecialHirelings: 4,
            Score: 9,
          },
          Constitution: {
            HitPoints: -1,
            RaiseDeadSurvivalChance: '50%',
            Score: 5,
          },
          Dexterity: {
            ArmorClass: 1,
            MissilesToHit: 1,
            Score: 13,
          },
          Gold: 140,
          HitPoints: 1,
          Intelligence: {
            MaxAdditionalLanguages: 4,
            MaxSpellLevel: 8,
            NewSpellUnderstandingChance: 75,
            Score: 15,
            SpellsPerLevel: '6/10',
          },
          Strength: {
            Carry: 5,
            Damage: 0,
            Doors: '1-2',
            Score: 9,
            ToHit: 0,
          },
          Wisdom: {
            Score: 9,
          },
        },
      }
      state.setCharacter(DEFAULT_INVENTORY_ID, characterOptions)

      let inventory = state.getInventory(DEFAULT_INVENTORY_ID)
      expect(inventory.character).toMatchSnapshot()

      state.removeCharacter(DEFAULT_INVENTORY_ID)
      inventory = state.getInventory(DEFAULT_INVENTORY_ID)
      expect(inventory.character).toBeNull()
    })
  })

  describe('inventories', () => {
    it('should create a new inventory', () => {
      const newInventory = State.getNewInventory('newId', 'New Inventory')
      expect(newInventory).toEqual({
        character: null,
        id: 'newId',
        items: { ...DEFAULT_INVENTORY_ITEMS },
        name: 'New Inventory',
      })
    })

    it('should set and get inventory', () => {
      const newInventory = State.getNewInventory('newId', 'New Inventory')
      state.setInventory('newId', newInventory)
      const inventory = state.getInventory('newId')
      expect(inventory).toEqual(newInventory)
    })

    it('should remove inventory', () => {
      const newInventory = State.getNewInventory('newId', 'New Inventory')
      state.setInventory('newId', newInventory)
      state.removeInventory('newId')
      const inventory = state.getInventory('newId')
      expect(inventory).toBeUndefined()
    })

    it('should remove inventory', () => {
      const newInventory = State.getNewInventory('newId', 'New Inventory')
      state.setInventory('newId', newInventory)
      state.removeInventory('newId')
      const inventory = state.getInventory('newId')
      expect(inventory).toBeUndefined()
    })

    it('should allow removing default inventory', () => {
      state.removeInventory(DEFAULT_INVENTORY_ID)
      expect(state.getInventory(DEFAULT_INVENTORY_ID)).toBeUndefined()
      expect(state.getInventories()).toEqual([])
    })
  })
})
