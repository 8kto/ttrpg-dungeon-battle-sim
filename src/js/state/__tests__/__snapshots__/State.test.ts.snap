// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State character should set and remove character in inventory 1`] = `
{
  "characterClass": "MagicUser",
  "spells": {
    "Charm Person": {
      "level": 1,
      "name": "Charm Person",
    },
    "Detect Magic": {
      "level": 1,
      "name": "Detect Magic",
    },
    "Hold Portal": {
      "level": 1,
      "name": "Hold Portal",
    },
    "Light": {
      "level": 1,
      "name": "Light",
    },
    "Protection from Evil": {
      "level": 1,
      "name": "Protection from Evil",
    },
    "Read Languages": {
      "level": 1,
      "name": "Read Languages",
    },
    "Shield": {
      "level": 1,
      "name": "Shield",
    },
    "Sleep": {
      "level": 1,
      "name": "Sleep",
    },
  },
  "stats": {
    "Charisma": {
      "MaxNumberOfSpecialHirelings": 4,
      "Score": 9,
    },
    "Constitution": {
      "HitPoints": -1,
      "RaiseDeadSurvivalChance": "50%",
      "Score": 5,
    },
    "Dexterity": {
      "ArmorClass": 1,
      "MissilesToHit": 1,
      "Score": 13,
    },
    "Gold": 140,
    "HitPoints": 1,
    "Intelligence": {
      "MaxAdditionalLanguages": 4,
      "MaxSpellLevel": 8,
      "NewSpellUnderstandingChance": 75,
      "Score": 15,
      "SpellsPerLevel": "6/10",
    },
    "Strength": {
      "Carry": 5,
      "Damage": 0,
      "Doors": "1-2",
      "Score": 9,
      "ToHit": 0,
    },
    "Wisdom": {
      "Score": 9,
    },
  },
}
`;

exports[`State default state should initialize with default inventory 1`] = `
[
  {
    "character": null,
    "id": "MainCharacter",
    "items": {
      "Basic accessories": {
        "cost": 0,
        "name": "Basic accessories",
        "quantity": 1,
        "weight": 8,
      },
    },
    "name": "Main Character",
  },
]
`;
